DatePicker component added.

tumentumurchudur opened the pull request:
DatePicker component added.

cerinman commented:
completes https://github.com/mxenabled/mx-react-components/issues/26

bsbeeks commented:
Please provide a writeup similar to https://github.com/mxenabled/mx-react-components/pull/3

bsbeeks commented:
Also, new components should use ES6 classes. See the other components for reference.

tumentumurchudur commented:
Sounds good. I will rewrite the component using ES6 classes and add some documentation.

tumentumurchudur commented:
DatePicker has been rewritten using ES6 classes. Please see the first draft of my documentation for the component and its parameters.

To use:
  const { DatePicker } = require('mx-react-components');
  <DatePicker onDateSelect={callbackFunction} selectedDate={Currently selected date}/>

There are no required parameters for this component but to make it useful, you need to provide a callback function that is called when a date is selected on the calendar as well as the parameter selectedDate to keep track of the currently selected date as a state in your component. 

Parameter Descriptions:
- calendarColumns (number, 7): A number of days to show in each column on the calendar. 
- closeOnDateSelect (boolean, false): Close the calendar automatically once a date is selected.
- divideCells (boolean, false):  Show border around each day on the calendar.
- fontSize (string, '1em'): Font size of the text and days.
- format (string, 'YYYY-MM-DD'): Date format of the calendar.
- locale (string, 'en'): Calendar locale
- minimumDate (string, null): Minimum date you can select on the calendar. There is no minimum date when it is null.
- onDateSelect (func, {}): Callback function on a date select.
- scrimStyle (object, {}): A style object that changes the styles of scrim. 
- selectedDate (number): Currently selected date on the calendar.
- useInputForSelectedDate (boolean, true): Allows the user to directly type in a date.
- useScrim (boolean, false): Show scrim

<img width="335" alt="screen shot 2015-10-19 at 4 08 43 pm" src="https://cloud.githubusercontent.com/assets/9920303/10592800/b4a40cda-767b-11e5-888d-dd0fc7599a5b.png">

bsbeeks commented on package.json:
```diff
@@ -41,6 +41,7 @@
     "eslint-plugin-react": "^3.4.2",
     "eslint-watch": "^2.1.1",
     "live-server": "^0.8.1",
+    "moment": "^2.10.3",
```
If the component uses moment, then this should be in `dependencies`.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+
+    this.setState({
+      currentDate
+    });
```
Should clicking the next/prev buttons actually change the selected date? I think a better UX would be to change the month, but not select a date. Leaving it up to the user to actually select the date they want.

bsbeeks replied:
I just pulled down and started using the component. It appears that this is functioning as I described above. The confusion came from the naming of the property.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+
+  render () {
+    const selectedDate = moment.unix(this.props.selectedDate).locale(this.props.locale);
+    const currentDate = this.state.currentDate ? this.state.currentDate.locale(this.props.locale) : selectedDate;
```
We are defining these two variables the same way a few times throughout the file. Perhaps we should be doing this in `constructor` or `componentDidMount`?

tumentumurchudur replied:
Yes, they are little repetitive. We could declare them in constructor as this.selectedDate and this.currentDate but it won't reduce the # of lines of code and these variables would become global. Dealing with global variables maybe little risky. Whereas, const variables are block-scoped and only used inside that function and destroyed immediately.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+        fontSize: '1em',
+        margin: '1px'
+      }
+    };
```
Could we move this `styles` object outside of the React class? In the places we are using `props` to set styles, we can modify the styles object or use Radium to our advantage and do something like
```
style={[styles.calendarHeader, { borderBottomStyle: this.props.divideCells ? 'solid' : 'none'}]}
```
This would also make it so we don't have to pass the `styles` variable to the `_renderMonthlyTable` method.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+          />
+          <Icon
+            onClick={this._toggleCalendar.bind(this)}
+            size='1.75em'
```
We try to avoid using `em`. Could we change this to a pixel size?

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+          onClick={this._toggleCalendar.bind(this)}
+          style={styles.selectedDate}
+        >
+          {moment.unix(this.props.selectedDate).format(this.props.format)}
```
I've noticed that we are relying on the user to pass in the `selectedDate` as a prop. To follow suit with our other components, we should make sure the component works independent of props being passed in. In this specific case, we should be keeping track of the selected date as a property on the state. Whenever it is changed we can expose it back to the user in a method prop, then they can do what they want with it.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+          {this.props.title}
+        </div>
+      );
+    } else {
```
You should be able to remove this `else` statement all together. If there is no title, then we don't return anything.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+        float: 'left',
+        paddingBottom: 100/(this.props.calendarColumns + 2) + '%',
+        position: 'relative',
+        width: 100/(this.props.calendarColumns + 1) + 1 + '%'
```
Could we somehow use css flexbox to simplify the day columns and week rows?

tumentumurchudur replied:
When I change the number of calendarColumns, the calendar does not look right. I think we should remove the prop calendarColumns all together.  Most calendars have a 7 day column anyway.

bsbeeks replied:
I agree, let's hard code it to 7 days per row

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+DatePicker.propTypes = {
+  calendarColumns: React.PropTypes.number,
+  closeOnDateSelect: React.PropTypes.bool,
+  divideCells: React.PropTypes.bool,
```
A better name for this might be, `showDayBorders`?

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+    while (startDate.isBefore(endDate)) {
+      const isCurrentMonth = startDate.month() === currentDate.month();
+      const isCurrentDay = startDate.format(this.props.format) === selectedDate.format(this.props.format);
+      const day = startDate.isBefore(minimumDate) ? (
```
This should probably be a standard if/else conditional instead of a ternary.

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,454 @@
+    }
+
+    //Divide calendar columns/Rows
+    if (this.props.divideCells) {
```
We are already looping through the days above in the `while` statement. Is there a way we can move this logic into that statement? At first glance it looks like we could use Radium to apply the appropriate styles.

bsbeeks commented:
This is looking good and we're definitely on the right track. Other things that need to be addressed are:
- Add CSS classes on all the elements that would make sense to style
- Make sure eslint is passing
- Make sure it's accessible (Here is an example: http://accessibility.athena-ict.com/aria/examples/datepicker2.shtml)

tumentumurchudur commented:
- Minor details have been modified as discussed
- Ran eslint and passed
- Touched up the UI and made it look as close as possible to the expected image above
- What css classes do you want added back? 

Now working on making the component accessible
<img width="293" alt="screen shot 2015-10-20 at 3 54 01 pm" src="https://cloud.githubusercontent.com/assets/9920303/10622175/e3538746-7742-11e5-857f-ee0faaa06f17.png">

tumentumurchudur commented:
The following changes have been made:
- circle background when a date is selected
- added a caret
- used font from StyleConstants
- removed space between the calendar and input
- used shadow from StyleConstants
- added props for default selected date and caret
- centered selected date inside input

bsbeeks commented on src/components/DatePicker.js:
```diff
@@ -0,0 +1,462 @@
+  minimumDate: React.PropTypes.string,
+  onDateSelect: React.PropTypes.func,
+  scrimStyle: React.PropTypes.object,
+  selectedDate: React.PropTypes.string,
```
This should be named `defaultDate` since once a user starts interacting with the date picker we stop using it. It will also help us differentiate between the `selectedDate` parameter on the state.

bsbeeks commented:
We're getting closer. A couple things I'm noticing:
- The font is still a serif font. It should be a sans-serif font using `StyleConstants. FontFamily`
- The padding and height of the selected date area doesn't match the expected design
- The circle indicator is still slightly oval, it should be a perfect circle
- The font color of the selected date looks correct, but the calendar font color appears darker
- The caret color in the selected date area is too dark

tumentumurchudur commented:
Modified font, padding, height, font color and faded carets.
<img width="377" alt="screen shot 2015-10-22 at 10 53 45 am" src="https://cloud.githubusercontent.com/assets/9920303/10672184/b27859c0-78ab-11e5-92fc-1cd64f36d696.png">

bsbeeks commented:
Looks like Travis is failing due to a trailing comma. Once that is fixed, this will be good to go.

bsbeeks commented on src/constants/Style.js:
```diff
@@ -19,5 +19,6 @@ module.exports = {
   FontSize: '13px',
-  FontFamily: 'Helvetica, Arial, sans-serif'
+  FontFamily: 'Helvetica, Arial, sans-serif',
+  BoxShadow: '0 30px 30px 10px rgba(0,0,0,0.1)',
```
the culprit

tumentumurchudur commented:
Yup, it compiles now. Thanks
